revamp grep port due to 35915ffebe and adjust its dependents

ylluminarious opened the pull request:
Made as per the discussion on the mailing list about how 35915ffebe breaks things. The beginning of said discussion can be found [here.](https://lists.macports.org/pipermail/macports-dev/2018-June/039101.html) A description of what's going on in these changes can be found in my commit message body.

The only problem that I can see are a couple of particular ports which are actually scripts instead of binaries. They are `p5-config-autoconf` and `vcs`. These ports specify a dependency on `port:grep` specifically so I assume that they rely on GNU grep during runtime. I guess they need to be modified to look for `ggrep`. Should this be done via `post-destroot` and `reinplace` or should something else be done here?

#### Description

<!-- Note: it is best make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk '{print $NF}' | tr '\n' ' ')"
-->
macOS 10.13.5
Xcode 9.4.1

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

Schamschula commented:
Given that we are now installing ggrep and friends, rather than grep, and then linking to binaries w/o the prefix, it might be wise to rename the port to ggrep, like gawk and gsed.

lbschenkel commented:
It is a small thing, but I would prefer if `dict` was modified as follows:
```
configure.env       GREP=${prefix}/libexec/gnubin/grep \
                    LIBTOOL=${prefix}/bin/glibtool
```

raimue commented:
LGTM. I would also add a note to the grep port to notify users of the change.

michaelld commented:
For my understanding: By removing "grep" from the dependency list for most of these ports, you're saying that these ports don't actually require "grep" (the executable) ... at least in any useful capacity ... yes?

As an example: "less"'s configure script checks for "grep", and actually uses "grep" directly in it (not $GREP or the like) ... but, that's the only place in the whole code that "grep" is actually used; hence, "less" doesn't really require port:grep & removing that dependency is OK (and, really, wise).

By the way: I do think this is a good change; I'm just coming up to speed on it.

michaelld commented:
I have a vague recollection that one reason some ports use "grep" (etc) is that the MacPorts version played nicely with locale settings, whereas the Xcode version didn't -- mostly on now-older OSXs I'd guess. We should make sure to test thoroughly not just on recent Xcode / OSX but on older versions too. I can & will do some in this regard; hopefully @kencu can too.

ylluminarious commented:
@Schamschula 

> Given that we are now installing ggrep and friends, rather than grep, and then linking to binaries w/o the prefix, it might be wise to rename the port to ggrep, like gawk and gsed.

Sounds good to me. I think I'll let other maintainers chime in to decide whether we should rename it or not.

---

@lbschenkel 

> It is a small thing, but I would prefer if dict was modified as follows:

```
configure.env       GREP=${prefix}/libexec/gnubin/grep \
                    LIBTOOL=${prefix}/bin/glibtool
```

Okay, I can try to do that and I'll push it here.

---

@raimue 

> LGTM.

Awesome, thank you!

>  I would also add a note to the grep port to notify users of the change.

Okay, I will do that. Should this note stay there for the indefinite future or should we remove it after a time?

---

@michaelld Thank you for your input. I'm glad that the changes look good to you. I also would very much appreciate any testing that could be done on older systems. I do not readily have access to such systems right now so 10.12 - 10.13 is the best that I can do.

raimue commented:
`bin:grep:grep` did not force installation of the grep port. Now that the grep port no longer provides `grep`, the dependency would not be fulfilled by the grep port anyway. If some port cannot work with `/usr/bin/grep`, the dependency on GNU grep had to be explicit before this change already (`port:grep`). The latter now needs to be convinced to use ggrep.

michaelld commented:
OK yes that makes sense now. I say "go for it" with the changes requested & we'll see how the cookies crumble.

Schamschula commented:
Agreed. Let's go with it and fix things as needed.

pmetzger commented:
Okay, so is this now considered good to commit?

ylluminarious commented:
@raimue I added some notes to the grep port in my new commit. @lbschenkel I've also changed the `dict` port as you specified. Please let me know what you guys think.

Furthermore, if anyone has any advice on what should be done with `vcs` and `p5-config-autoconf`, that would be helpful. Once again, these two ports are scripts which specify a hard dependency on `port:grep`. I've looked around at each of them, and I've tried to see if they can be coerced to use the `g` prefix, but I'm afraid that I may cause further breakage if I muck around in these ports (which I don't even use).

Lastly, does anyone have any thoughts on renaming this whole port to `ggrep` as per @Schamschula's suggestion above?

lbschenkel commented:
+1 for the `dict` change

pmetzger commented:
Is this now good to commit?

ylluminarious commented:
@pmetzger I'd say so if no one else has any further comments or suggestions.

michaelld commented:
LGTM

pmetzger commented:
> Given that we are now installing ggrep and friends, rather than grep, and then linking to binaries w/o the prefix, it might be wise to rename the port to ggrep, like gawk and gsed.

So what about this last detail? Should we be renaming it at the same time, or should that happen later?

Schamschula commented:
I'll do the renaming step separately. Please go ahead with the merge.

pmetzger commented:
Merged. Thanks to all involved, and of course to @ylluminarious.

ylluminarious commented:
@pmetzger Thank you! And thanks to everyone else involved!
